storage/remote/read_test.go

```diff
@@ -494,3 +533,266 @@ func TestSampleAndChunkQueryableClient(t *testing.T) {
 		})
 	}
 }
+
+func TestReadMultiple(t *testing.T) {
+	m := &mockedRemoteClient{
+		store: []*prompb.TimeSeries{
+			{Labels: []prompb.Label{{Name: "job", Value: "prometheus"}}},
+			{Labels: []prompb.Label{{Name: "job", Value: "node_exporter"}}},
+			{Labels: []prompb.Label{{Name: "job", Value: "cadvisor"}, {Name: "region", Value: "us"}}},
+			{Labels: []prompb.Label{{Name: "instance", Value: "localhost:9090"}}},
+		},
+		b: labels.NewScratchBuilder(0),
+	}
+
+	testCases := []struct {
+		name            string
+		queries         []*prompb.Query
+		expectedResults []labels.Labels
+	}{
+		{
+			name: "single query",
+			queries: []*prompb.Query{
+				{
+					StartTimestampMs: 1000,
+					EndTimestampMs:   2000,
+					Matchers: []*prompb.LabelMatcher{
+						{Type: prompb.LabelMatcher_EQ, Name: "job", Value: "prometheus"},
+					},
+				},
+			},
+			expectedResults: []labels.Labels{
+				labels.FromStrings("job", "prometheus"),
+			},
+		},
+		{
+			name: "multiple queries - different matchers",
+			queries: []*prompb.Query{
+				{
+					StartTimestampMs: 1000,
+					EndTimestampMs:   2000,
+					Matchers: []*prompb.LabelMatcher{
+						{Type: prompb.LabelMatcher_EQ, Name: "job", Value: "prometheus"},
+					},
+				},
+				{
+					StartTimestampMs: 1500,
+					EndTimestampMs:   2500,
+					Matchers: []*prompb.LabelMatcher{
+						{Type: prompb.LabelMatcher_EQ, Name: "job", Value: "node_exporter"},
+					},
+				},
+			},
+			expectedResults: []labels.Labels{
+				labels.FromStrings("job", "node_exporter"),
+				labels.FromStrings("job", "prometheus"),
+			},
+		},
+		{
+			name: "multiple queries - overlapping results",
+			queries: []*prompb.Query{
+				{
+					StartTimestampMs: 1000,
+					EndTimestampMs:   2000,
+					Matchers: []*prompb.LabelMatcher{
+						{Type: prompb.LabelMatcher_RE, Name: "job", Value: "prometheus|node_exporter"},
+					},
+				},
+				{
+					StartTimestampMs: 1500,
+					EndTimestampMs:   2500,
+					Matchers: []*prompb.LabelMatcher{
+						{Type: prompb.LabelMatcher_EQ, Name: "region", Value: "us"},
+					},
+				},
+			},
+			expectedResults: []labels.Labels{
+				labels.FromStrings("job", "cadvisor", "region", "us"),
+				labels.FromStrings("job", "node_exporter"),
+				labels.FromStrings("job", "prometheus"),
+			},
+		},
```

Comment:
What about the case where the same set of matchers is used for overlapping time ranges (eg. `region="us"` for T=1 to T=5 and T=3 to T=8)

Reply:
added this in https://github.com/prometheus/prometheus/pull/16742/commits/1512f575c6f84980851ea926fbaaa00dfa3fd9bc